### PR TITLE
Remove Binder.CheckFeatureAvailability that uses useSiteDiagnostics

### LINKED ISSUE #25028: Fix `GetStackAllocConversion` and remove `Binder.CheckFeatureAvailability` method
During overload resolution or conversion classification, we use use-site diagnostics to report problems with symbols that might affect the correctness of result. 
`GetStackAllocConversion` should not be putting language-version-related diagnostics as use-site diagnostics since there was nothing wrong with symbols. Also, we don't want language version to affect the result of the binding (including conversion classification).

We should review this method's use of use-site diagnostics to comply with this guideline.

This was reported by @AlekseyTs in a PR that was also going to mis-use use-site diagnostics (https://github.com/dotnet/roslyn/pull/24913#discussion_r170335206).

```

        public override Conversion GetStackAllocConversion(BoundStackAllocArrayCreation sourceExpression, TypeSymbol destination, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
        {
...
                            Binder.CheckFeatureAvailability(sourceExpression.Syntax, MessageID.IDS_FeatureRefStructs, ref useSiteDiagnostics);
                            Binder.GetWellKnownTypeMember(_binder.Compilation, WellKnownMember.System_Span_T__ctor, out DiagnosticInfo memberDiagnosticInfo);
                            HashSetExtensions.InitializeAndAdd(ref useSiteDiagnostics, memberDiagnosticInfo);

                            return Conversion.MakeStackAllocToSpanType(spanConversion);
                        }
...
        }
```

Tagging @VSadov @OmarTawfik

### PR DESCRIPTION
**Pull request by OmarTawfik**
Fixes #25028

cc @jcouv @AlekseyTs 
@dotnet/roslyn-compiler for review

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs [hunk 1]
```diff
@@ -312,8 +312,7 @@ public override Conversion GetStackAllocConversion(BoundStackAllocArrayCreation
 
                         if (spanConversion.Exists)
                         {
-                            // Report errors if Span ctor is missing, or using an older C# version
-                            Binder.CheckFeatureAvailability(sourceExpression.Syntax, MessageID.IDS_FeatureRefStructs, ref useSiteDiagnostics);
+                            // Report errors if Span ctor is missing
                             Binder.GetWellKnownTypeMember(_binder.Compilation, WellKnownMember.System_Span_T__ctor, out DiagnosticInfo memberDiagnosticInfo);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs]: >Binder.GetWellKnownTypeMember(_binder.Compilation, WellKnownMember.System_Span_T__ctor, out DiagnosticInfo memberDiagnosticInfo); [](start = 28, length = 129)

I don't think we should be performing this check here either. Whether the constructor is good has no effect on the result of conversion classification, we still return the same Conversion value. Instead, the use-site diagnostics for the constructor should be reported where we actually try to use it. I see two other places where we look for the same symbol:
```
  src\Compilers\CSharp\Portable\Binder\Binder_Expressions.cs(3128):                GetWellKnownTypeMember(Compilation, WellKnownMember.System_Span_T__ctor, diagnostics, syntax: node);
  src\Compilers\CSharp\Portable\Lowering\LocalRewriter\LocalRewriter_StackAlloc.cs(42):                var spanCtor = (MethodSymbol)_compilation.GetWellKnownTypeMember(WellKnownMember.System_Span_T__ctor).SymbolAsMember(spanType);
```

The first one also doesn't use the symbol, simply checks its "goodness". I think we should remove that check as well, and the goodness should only be checked in LocalRewriter, which does not do this at the moment. See how LocalRewriter.TryGetWellKnownTypeMember helper is used elsewhere in the rewriter.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1).

**Comment by AlekseyTs** [comment]
>                     _binder.ReportUnsafeIfNotAllowed(sourceExpression.Syntax.Location, ref useSiteDiagnostics);

This is another instance of misuse of use ```useSiteDiagnostics```, please fix it as well.

---
Refers to: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:301 in 867f475. [](commit_id = 867f47561080de659dcb455b6e347f4578b695ea, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StackAlloc.cs [hunk 2]
```diff
@@ -39,14 +39,27 @@ public override BoundNode VisitStackAllocArrayCreation(BoundStackAllocArrayCreat
                 var stackSize = RewriteStackAllocCountToSize(countTemp, elementType);
                 stackAllocNode = new BoundConvertedStackAllocExpression(stackAllocNode.Syntax, elementType, stackSize, spanType);
 
-                var spanCtor = (MethodSymbol)_compilation.GetWellKnownTypeMember(WellKnownMember.System_Span_T__ctor).SymbolAsMember(spanType);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StackAlloc.cs]: >.SymbolAsMember(spanType) [](start = 117, length = 25)

It looks like you lost this part during refactoring

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 3)

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs [hunk 3]
```diff
@@ -291,15 +291,11 @@ public override Conversion GetStackAllocConversion(BoundStackAllocArrayCreation
...
 
-                var pointerConversion = default(Conversion);
                 var sourceAsPointer = new PointerTypeSymbol(sourceExpression.ElementType);
-                pointerConversion = ClassifyImplicitConversionFromType(sourceAsPointer, destination, ref useSiteDiagnostics);
+                var pointerConversion = ClassifyImplicitConversionFromType(sourceAsPointer, destination, ref useSiteDiagnostics);
 
                 if (pointerConversion.IsValid)
                 {
-                    // Report unsafe errors
-                    _binder.ReportUnsafeIfNotAllowed(sourceExpression.Syntax.Location, ref useSiteDiagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs]: >ReportUnsafeIfNotAllowed [](start = 28, length = 24)

It looks like this overload is no longer used, please remove it

**Comment by jcouv** [comment]
Unrelated to this PR: a comment from Aleksey in another PR made me think of this: what if you create a custom `Span<T>` type with some constraints on `T`. Do we properly check and report on such constraints?
See `BindStackAllocArrayCreationExpression` line 3130 (`type = spanType.Construct(elementType);`).

**Comment by jcouv** [comment]
@jaredpar for ask-mode approval for 15.7. Thanks

**Comment by OmarTawfik** [comment]
@jcouv can you please create an issue with that comment from the other PR so that we don't drop it?
We should be adding tests for that edge case anyways.

**Comment by jcouv** [comment]
Filed https://github.com/dotnet/roslyn/issues/25086

**Comment by jaredpar** [comment]
approved
